onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -127,8 +127,43 @@ static bool IsRelevantOutput(const Node* node, const NodeArg* output) {
   return true;
 }
 
+// Check whether the node is a cast operation from float16/float to the specified data_type.
+static bool IsCastTo(const Node* node, TensorProto_DataType data_type) {
+  if (node && node->OpType() == "Cast") {
+    const NodeAttributes& attributes = node->GetAttributes();
+    const auto attr_hit = attributes.find("to");
+    ORT_ENFORCE(attr_hit != attributes.end(), "Node: ", node->Name(),
+                " is a Cast node and it must have 'to' attribute set.");
+    const NodeArg* input = node->InputDefs()[0];
+    auto input_data_type = static_cast<TensorProto_DataType>(input->TypeAsProto()->tensor_type().elem_type());
+    // Allow cast nodes with same input and output type float/float16 to eliminate such casts.
+    return (input_data_type == TensorProto::FLOAT16 || input_data_type == TensorProto::FLOAT) &&
+           attr_hit->second.i() == static_cast<int64_t>(data_type);
+  }
+  return false;
+}
+
+// when node is softmax, and its input comes from a cast-to-fp32 node, and its output is only consumed by one cast-to-fp16 node, then we can treat it as a fp16-allowed op,
+// as ort's softmax implementation already does the necessary cast logic, for example do reduce sum at fp32
+static bool SoftmaxCanBeFP16(const Node& node) {
+  if (node.OpType() != "Softmax")
+    return false;
+  // 1. input comes from a cast-to-fp32 node
+  const Node* input_node = graph_utils::GetInputNode(node, 0);
+  if (!(input_node && IsCastTo(input_node, TensorProto::FLOAT)))
+    return false;
+  // 2. output is consumed by a cast-to-fp16 node ONLY
+  if (node.GetOutputEdgesCount() != 1)
+    return false;
+  const Node* output_node = &(*node.OutputNodesBegin());
+  if (!(output_node && IsCastTo(output_node, TensorProto::FLOAT16)))
+    return false;
+
+  return true;
+}
+
 // Check whether the given opcode is fp16 allowed for the given level of optimization.
-static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_level0_ops) {
+static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_level0_ops, const Node* node = nullptr) {
```

Comment:
why not make `const Node* node` be a required param, and remove `const std::string& op_type`?